ion-schema/src/lib.rs

```diff
@@ -175,37 +225,63 @@ pub struct UserReservedFields {
 
 impl UserReservedFields {
     /// Parse use reserved fields inside a [Struct]
-    pub(crate) fn from_ion_elements(user_reserved_fields: &Struct) -> Self {
-        Self {
+    pub(crate) fn from_ion_elements(user_reserved_fields: &Struct) -> IonSchemaResult<Self> {
+        if user_reserved_fields.fields().any(|(f, v)| {
+            f.text() != Some("schema_header")
+                && f.text() != Some("schema_footer")
+                && f.text() != Some("type")
+        }) {
+            return invalid_schema_error(
+               "User reserved fields can only have schema_header, schema_footer or type as the field names",
+           );
+        }
+        Ok(Self {
             schema_header_fields: UserReservedFields::field_names_from_ion_elements(
                 "schema_header",
                 user_reserved_fields,
-            ),
+            )?,
             schema_footer_fields: UserReservedFields::field_names_from_ion_elements(
                 "schema_footer",
                 user_reserved_fields,
-            ),
+            )?,
             type_fields: UserReservedFields::field_names_from_ion_elements(
                 "type",
                 user_reserved_fields,
-            ),
-        }
+            )?,
+        })
     }
 
     fn field_names_from_ion_elements(
         user_reserved_fields_type: &str,
         user_reserved_fields: &Struct,
-    ) -> Vec<String> {
+    ) -> IonSchemaResult<Vec<String>> {
         let user_reserved_elements: Vec<&Element> = user_reserved_fields
             .get(user_reserved_fields_type)
             .and_then(|it| it.as_sequence().map(|s| s.elements().collect()))
-            .unwrap_or(vec![]);
+            .ok_or(invalid_schema_error_raw(
+                "User reserved fields mut be non null",
+            ))?;
 
-        user_reserved_elements
+        let user_reserved_fields = user_reserved_elements
             .iter()
-            .map(|e| e.as_string().map(|s| s.to_owned()))
+            .filter(|e| e.annotations().is_empty() && !e.is_null())
+            .map(|e| e.as_text().map(|s| s.to_owned()))
             .collect::<Option<Vec<String>>>()
-            .unwrap_or(vec![])
+            .unwrap_or(vec![]);
+
+        if user_reserved_fields.len() != user_reserved_elements.len() {
+            return invalid_schema_error("User reserved fields mut be unannotated");
+        }
+
+        if user_reserved_fields.iter().any(|f| {
+            is_reserved_keyword_isl_version_marker(f) || ISL_2_0_KEYWORDS.contains(&f.as_str())
```

Comment:
Small optimization opportunity since `ISL_2_0_KEYWORDS` is in alphabetical order: use [`binary_search`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search) instead of `contains`.